File: ts/test/grammar-immutable.test.js
/* Copyright (c) 2026 Richard Rodger and other contributors, MIT License */
'use strict'

// `grammar()` must not modify the spec it is given.
//
// It used to. Normalisation is destructive by design — `normalt` rewrites
// `g` into a sorted array and nulls an empty `s`, and `resolveFunctionRef`
// swaps each `@name$` string for the resolved function — and all of it ran
// directly on the caller's object. A declarative grammar came out full of
// functions, so serialising *after* installing produced a spec with every
// action missing. That spec still installed and still parsed; it just
// returned `undefined`. The failure was silent, which is what made it
// expensive.
//
// Go never had the bug: `Grammar` leaves the caller's *GrammarSpec alone.
// These tests pin the TS side to the same contract.

const { describe, it } = require('node:test')
const assert = require('node:assert')

const { Tabnas } = require('..')

// A grammar with every ref-carrying shape: a ref array (`a`), a single ref
// string (`a`), an empty `s`, and an unsorted comma-string `g`.
const makeSpec = () => ({
  options: {
    fixed: { token: { '#EQ': '=' } },
    rule: { start: 'pair' },
  },
  rule: {
    pair: {
      open: [{ s: '#TX #EQ', p: 'val', a: ['@object$', '@key$'], g: 'zulu,alpha' }],
      close: [{ a: '@setval$' }],
    },
    val: {
      open: [{ s: '#NR', a: '@value$' }],
      close: [{}],
    },
  },
})

describe('grammar-immutable', () => {

  it('leaves the caller spec byte-identical', () => {
    const spec = makeSpec()
    const before = JSON.stringify(spec)

    new Tabnas().grammar(spec)

    assert.equal(JSON.stringify(spec), before,
      'grammar() modified the spec it was given')
  })

  it('keeps ref strings as strings', () => {
    const spec = makeSpec()
    new Tabnas().grammar(spec)

    assert.deepEqual(spec.rule.pair.open[0].a, ['@object$', '@key$'])
    assert.equal(spec.rule.pair.close[0].a, '@setval$')
    assert.equal(typeof spec.rule.val.open[0].a, 'string')
  })

  it('leaves g and s untouched', () => {
    const spec = makeSpec()
    new Tabnas().grammar(spec)

    // normalt sorts g into an array and nulls an empty s — on its own copy.
    assert.equal(spec.rule.pair.open[0].g, 'zulu,alpha')
    assert.deepEqual(spec.rule.val.close[0], {})
  })

  it('installs the same spec value twice with the same result', () => {
    const spec = makeSpec()

    const first = new Tabnas()
    first.grammar(spec)
    assert.deepEqual(first.parse('port = 8080'), { port: 8080 })

    // The bug: the second install saw a spec whose actions had been replaced
    // by resolved functions on the first pass, and quietly produced nothing.
    const second = new Tabnas()
    second.grammar(spec)
    assert.deepEqual(second.parse('port = 8080'), { port: 8080 })
  })

  it('round-trips through JSON after installing', () => {
    const spec = makeSpec()

    const tn = new Tabnas()
    tn.grammar(spec)
    assert.deepEqual(tn.parse('port = 8080'), { port: 8080 })

    // Serialise AFTER install — the order the docs used to warn against.
    const wire = JSON.stringify(spec)
    assert.ok(wire.includes('@object$'), 'refs were lost from the spec')

    const elsewhere = new Tabnas()
    elsewhere.grammar(JSON.parse(wire))
    assert.deepEqual(elsewhere.parse('port = 8080'), { port: 8080 })
  })

  it('still accepts real functions, by reference', () => {
    // The clone copies plain objects and arrays but must pass functions
    // through untouched — a spec may carry actual actions, not just refs.
    let called = 0
    const action = () => { called++ }

    const spec = makeSpec()
    spec.rule.val.open[0].a = action

    const tn = new Tabnas()
    tn.grammar(spec)
    tn.parse('port = 8080')

    assert.equal(spec.rule.val.open[0].a, action, 'function was copied, not referenced')
    assert.ok(0 < called, 'function action did not run')
  })

})

File: go/grammar_immutable_test.go
// Copyright (c) 2013-2026 Richard Rodger, MIT License

package tabnas

// Grammar must not modify the *GrammarSpec it is given.
//
// Go has always behaved this way, but nothing pinned it. The TS runtime
// did the opposite until 2026-08-07: normalisation ran directly on the
// caller's object, so `@name$` ref strings were replaced by resolved
// functions and a declarative grammar came back full of code. Serialising
// after installing then produced a spec with every action missing, which
// still installed, still parsed, and returned nothing.
//
// The shared .tsv fixtures pin how grammars parse, not what the engine API
// does to its arguments, which is exactly why that divergence survived so
// long. These tests close that hole on the Go side; the TS counterpart is
// ts/test/grammar-immutable.test.js.

import (
	"encoding/json"
	"testing"
)

func immutableSpec() *GrammarSpec {
	return &GrammarSpec{
		OptionsMap: map[string]any{
			"fixed": map[string]any{"token": map[string]any{"#EQ": "="}},
			"rule":  map[string]any{"start": "pair"},
		},
		Rule: map[string]*GrammarRuleSpec{
			"pair": {
				Open:  []*GrammarAltSpec{{S: "#TX #EQ", P: "val", A: []any{"@object$", "@key$"}, G: "zulu,alpha"}},
				Close: []*GrammarAltSpec{{A: "@setval$"}},
			},
			"val": {
				Open:  []*GrammarAltSpec{{S: "#NR", A: "@value$"}},
				Close: []*GrammarAltSpec{{}},
			},
		},
	}
}

func parsedPort(t *testing.T, j *Tabnas) string {
	t.Helper()
	out, err := j.Parse("port = 8080")
	if err != nil {
		t.Fatalf("parse: %v", err)
	}
	b, err := json.Marshal(out)
	if err != nil {
		t.Fatalf("marshal: %v", err)
	}
	return string(b)
}

// The caller's spec reads the same after Grammar as before it.
func TestGrammarLeavesCallerSpecIntact(t *testing.T) {
	gs := immutableSpec()

	openAlts, ok := gs.Rule["pair"].Open.([]*GrammarAltSpec)
	if !ok {
		t.Fatal("open alts not []*GrammarAltSpec")
	}
	alt := openAlts[0]

	j := Make()
	if err := j.Grammar(gs); err != nil {
		t.Fatalf("install: %v", err)
	}
	if got := parsedPort(t, j); `{"port":8080}` != got {
		t.Fatalf("parse = %s, want {\"port\":8080}", got)
	}

	// Ref strings stay strings — not resolved functions written back.
	refs, ok := alt.A.([]any)
	if !ok {
		t.Fatalf("A = %#v, want []any of ref strings", alt.A)
	}
	if 2 != len(refs) || "@object$" != refs[0] || "@key$" != refs[1] {
		t.Fatalf("A = %#v, want [@object$ @key$]", refs)
	}

	// G is not rewritten into a sorted slice on the caller's value.
	if "zulu,alpha" != alt.G {
		t.Fatalf("G = %q, want %q", alt.G, "zulu,alpha")
	}
}

// Installing the same spec value twice must give the same grammar both
// times. This is the case that failed silently in TS.
func TestGrammarInstallsSameSpecTwice(t *testing.T) {
	gs := immutableSpec()

	first := Make()
	if err := first.Grammar(gs); err != nil {
		t.Fatalf("first install: %v", err)
	}
	if got := parsedPort(t, first); `{"port":8080}` != got {
		t.Fatalf("first parse = %s", got)
	}

	second := Make()
	if err := second.Grammar(gs); err != nil {
		t.Fatalf("second install: %v", err)
	}
	if got := parsedPort(t, second); `{"port":8080}` != got {
		t.Fatalf("second parse = %s, want {\"port\":8080} — spec was consumed by the first install", got)
	}
}
